### ISSUE TITLE
Reciever cannot be recovered 

### ISSUE THREAD

**Issue by kereid** [aodn/aatams#251]
Email trail from James:

10 August 2015
Hi Guys,

Hope you are all well down in the snow!

I am trying to update some deployments and recoveries so that I can upload data.

I am running into a problem where I cannot recover a VR4 receiver 250579 because it tells me that I can’t have the deployment before the initialisation date. Strange thing is there is an initialisation date in the deployment information, which is impossible so it seems there is a bug.

I have attached two screen shots to this email to help explain the problem.

Kind Regards,

James van den Broek
![2015 08_aatams database screen shot 2](https://cloud.githubusercontent.com/assets/2906602/9213851/615fbe42-40da-11e5-8d65-b7848714a183.jpg)
![2015 08_aatams database screen shot2 2](https://cloud.githubusercontent.com/assets/2906602/9213850/615e70a0-40da-11e5-9e1d-f6cbde38feba.jpg)

Hi James,

The error you are seeing is correct for the database, the deployment date/time should be after the initialisation time so the initialisation date/time should be 27/05/15 03:36

Cheers,
Kate   

12 August 2015

Hi Kate,

The problem that we have is that if you look at the screen shot, there is an initialisation date and time (16/6/15 @ 01:51:53) linked to the last deployment on the 27/5/15. The initialisation date and time are entered at the Recovery stage which hasn’t been entered yet. The initialisation date and time that are mentioned above must be a faulty link in the database.

I ran into the problem whilst attempting to recover this receiver and enter the correct details.

Regards,
James

**Comment by kereid** [comment]
Email from James:

I’m currently looking at our unrecovered Narooma Line stations and they also all have a linked initialisation time…

**Comment by jkburges** [comment]
FWIW, I just tried creating a new deployment, and can confirm that the initialisation date is not set to anything (e.g. if you go in to the receiver screen, that field will be blank).  So, I'm not sure at what point it became set (to an incorrect value) w.r.t. bug report.

**Comment by astecken** [comment]
Hi guys, we urgently need to have this sorted as we can't recover/redeploy any of our data as it is. It basically became a problem after the newly introduced red outlines were introduced, I think.

**Comment by xhoenner** [comment]
@pblain can this become a VIT? It's a critical issue according to Andre.

**Comment by jkburges** [comment]
I'm a bit confused here.  Is the problem happening on the *receiver deployment* screen?  

The [screenshot above](https://cloud.githubusercontent.com/assets/2906602/9213850/615e70a0-40da-11e5-9e1d-f6cbde38feba.jpg) shows the *receiver recovery* screen, with init date and deployment date that are indeed invalid w.r.t. each other (the init date must be *before* or equal to the deployment date). 

So the fix is to enter an initialisation date that is before or equal to the deployment date.

Having said that, I think I'm missing something here.

**Comment by astecken** [comment]
The screenshot shows the receiver recovery screen to show that their is an initialisation time stated even though this is NOT possible (as the init time is entered during the recovery of a receiver and not at deployment). Where does this info come from? Must be cross-referenced to some other table here or something.
When we tried to recover the unit with its actual init time/recovery time, the DB gave us the msg that the times are colliding.

**Comment by pblain** [comment]
@xhoenner , we only have three developers at the moment.  Let's have a chat about this.

**Comment by xhoenner** [comment]
no worries @pblain, might require @jkburges to be present as I haven't fully grasped what the issue exactly is.

**Comment by jkburges** [comment]
@astecken @xhoenner heads up that we're now looking at this bug.

I have just tried to re-create it.  I *was* able to successfully enter a recovery with the following date/times:

Initialisation Date: 2015-05-27T03:37:00+0000
Recovery Date: 2015-07-30T03:45:33+0000

where 

* initialisation date is set equal to the deployment date (since the init date in screenshots above is *after* deployment - i.e. invalid)
* recovery date is taken from [the screenshot above](https://cloud.githubusercontent.com/assets/2906602/9213850/615e70a0-40da-11e5-9e1d-f6cbde38feba.jpg).

While I cannot explain (yet) how the init date above got there\*, I don't see why recovery data cannot be entered *provided that the init/deployment/recovery dates are valid* with respect to each other.  i.e. init <= deployment < recovery.

Am I still missing something? :smile:

\* possibly the init date is defaulting to the deployment date?

**Comment by astecken** [comment]
Ok, I will try to explain again:

The first screen shot with the newly introduced red outline shows an unrecovered receiver. When a receiver is deployed and not yet recovered (red outlined section), it is impossible to have a corresponding init time. The latter is entered by the researcher only AFTER the recovery of the receiver when the vrl files are downloaded. 

It is irrelevant if the init time is valid or invalid (or before or after the deployment time). It can simply not exist in the first place. And (from memory) the init time now showing came into place when the red outlines were introduced recently. So there must have been something going wrong in the background.

Is this clearer now? I am in the field at the moment but please try and give me a call if it's easier to discuss on the phone!

**Comment by jkburges** [comment]
I understand what you're saying (I think).

The init date is actually stored against the deployment in the DB, although as far as the user is concerned, it's entered on the recovery screen.  In the first screen shot, I'm not sure how that value got there, so that's a bug, no argument from me on that.

But, what I still don't understand is: provided valid dates are entered on the recovery screen, is there something stopping users from being able to enter recoveries?

**Comment by astecken** [comment]
Yes, the bug has to be fixed so that we can enter the actual init and recovery time :)

If I would make up any valid dates (that aren't real), I am sure there wouldn't be a problem. But the wrong init times that are visible in the first screenshot restrict us from entering actual dates and times that reflect data collected by the researcher.

**Comment by jkburges** [comment]
What are the actual dates and times, one example?

**Comment by jkburges** [comment]
To answer how the initialisation date ended up with a value in there in the first place, I just had a look at the code, and it used to be the case that the init date (as well as the deployment date) defaulted to the current date when a deployment was first created.  Now, users wouldn't have known (or cared) that this was happening, as the init date was never visible on the receiver screen (as it is now), and was set (to whatever they wanted) on the recovery screen - as it still is now.

This has changed - init date now does *not* have a default value set, which better reflects what the user is actually entering in. So, for any *newly entered* deployments, the init date column should remain blank.  

I'd still like to know an example set of init, deployment, recovery dates that you're having trouble with, @astecken, e.g. for the above deployment/receiver - what are the dates that you're *trying* to enter.

**Comment by xhoenner** [comment]
@astecken Jon needs your input to get that bug fixed. Could you provide him with an example of dates and times so that he can recreate that bug?

**Comment by astecken** [comment]
Enter 250579 in search box, select receiver from list, last entry is highlighted in red.

What does this newly highlighted box mean? Can I find information on that somewhere as it is not immediately clear for the user?

Initialisation date shouldn't exist yet as it is entered during recovery process but DB states: 2015-06-16T01:51:53+0000 (which is after deployment date and has never been entered (again: it couldn't have been entered as it is only entered during recovery process while downloading the vrl file from the receiver))
Deployment date is 27/5/15, 3:37
Recovery date 30/7/2015, 3:45:33

Due to the wrong init date/time, we cannot recover (and redeploy) receiver 250579 (and any other for that matter).

Thanks for sorting this out!

**Comment by jkburges** [comment]
Forget everything that you see in the system. What are the:

* init date
* deployment date
* recovery date

that you *want* to enter for this receiver?

**Comment by astecken** [comment]
Init date 
Deployment date is 27/5/15, 3:37
Recovery date 30/7/2015, 3:45:33

**Comment by astecken** [comment]
From what I can see from James/Phil's datasheets, it should say:

Init date 27/5/15, 3:38:07
Deployment date is 27/5/15, 3:39
Recovery date 30/7/2015, 3:45

**Comment by jkburges** [comment]
Ok thanks @astecken.  Given those dates, I now have confirmation as to why it was not possible for you to simply enter them in the recovery screen (as [I did above](https://github.com/aodn/aatams/issues/251#issuecomment-136586956) with *slightly* different dates).

Here's what we will do or have done to resolve this issue:

1. fix the bug where the init date was being initialised to be equal to the deployment date at the time the *deployment* is entered in the system (this is already done - but let us know if you still see it happening of course)
1. fix up any deployments which have an incorrectly initialised init date (such as this one).


To answer the question about the highlighted red box: it is to indicate that the dates in the init/deployment/recovery dates are valid with respect to each other.  It was brought in to help diagnose and correct the [issue to do with invalid deployments](https://github.com/aodn/aatams/issues/241).  It was released at the same time that we brought in proper validation of dates (which is part of the reason why you are now having this issue).

**Comment by jkburges** [comment]
Ok, I have now done 2) above.  For any deployments which don't have an associated recovery, the init date should now be not set.  Here is a list of the ones which were changed *just in case*: https://gist.github.com/jkburges/15f71ba71d19f4c7f33d

So, @astecken, you should now be able to enter deployment/recovery for the above receiver, and any others which were facing the same problem.

**Comment by astecken** [comment]
Thanks, Jon! Looks good.

**Comment by xhoenner** [comment]
@jkburges can this issue be closed then?

**Comment by xhoenner** [comment]
all good to close @jkburges?

**Comment by jkburges** [comment]
This issue *is* closed :confused:
